[Logs] Update the Monitor K8s tutorial to use Standalone Agent and the K8s integration (#3581)
elastic/observability-docs: docs/en/observability/monitor-k8s/monitor-k8s-add-integration.asciidoc

[discrete]
[[monitor-kubernetes-integration]]
== Part 1: Add and Configure the Kubernetes integration

> To start collecting logs and metrics from your Kubernetes clusters, first add the {integrations-docs}/kubernetes[Kubernetes integration] to your deployment and configure the metrics and logs you want to collect.

Reviewer: ```suggestion
To start collecting logs and metrics from your Kubernetes clusters, first add the {integrations-docs}/kubernetes[Kubernetes integration] to your policy and configure the metrics and logs you want to collect.
```

[discrete]
[[monitor-k8s-add-integration]]
=== Step 1: Add the Kubernetes integration to your deployment

> Follow these steps to add the Kubernetes integration to your deployment:

Reviewer: ```suggestion
Follow these steps to add the Kubernetes integration to your policy:
```

//add screenshots if necessary

. Select *Add integrations* from your deployment's homepage.
. Enter "Kubernetes" in the search bar, and select the *Kubernetes* integration.
. Click *Add Kubernetes* at the top of the Kubernetes integration page.
. Click *Add integration only (skip agent installation)* at the bottom of the Add integration page.

You're now on the *Add Kubernetes integration* page.
From here, you'll configure the integration to collect metrics and logs from the different Kubernetes components.
Continue to <<monitor-k8s-configure-integration, Step 2>> to configure your integration.

[discrete]
[[monitor-k8s-configure-integration]]
=== Step 2: Configure your Kubernetes integration

The Kubernetes integration can fetch metrics and logs from several components.
From the *Add Kubernetes integration* page, configure the options for the metrics and logs you want to collect.
The following sections outline the configuration options.

*Collect Metrics*

Collecting metrics about Kubernetes clusters and the workloads running on top of them is a key aspect of Kubernetes observability.

> However, collecting metrics from Kubernetes poses some challenges.

Reviewer: Consider removing whole line?

> You need to collect metrics about the resources running on "physical" machines as well as the containers and pods.

Reviewer: ```suggestion
You need to collect metrics about the resources running on underlying infrastructure (eg machines of k8s cluster), metrics for the whole cluster  as well as individual metrics for the containers and pods.
```

Specifically, you need to monitor the health and performance of:

* The hosts where Kubernetes components are running. Each host produces metrics
like CPU, memory, disk utilization, and disk and network I/O.

* Kubernetes containers, which produce their own set of metrics.

* The applications running as Kubernetes pods, such as application servers and
databases, each producing its own set of metrics.

> 

Reviewer: * Additional Kubernetes resources (like services, deploymets, cronjobs) are valuable assets of the whole infrastructure and produce their own set of metrics that need monitoring

> Instead of using multiple technologies to collect metrics, deploy the {agent} along with the Kubernetes integration to monitor all layers of your technology stack.

Reviewer: ```suggestion
 {agent} along with the Kubernetes integration provides a unified solution to monitor all layers of your {k8s} technology stack (Instead of using multiple technologies to collect metrics)
```

*Kubernetes ResourceQuota metrics*:: Monitor resource limits and current usage. Learn more at {integrations-docs}/kubernetes/kube-state-metrics#state_resourcequota[`kube-state-metrics` ResourceQuota metrics].
*Kubernetes Service metrics*:: Monitor service configuration, accessibility, and network integration. Learn more at {integrations-docs}/kubernetes/kube-state-metrics#state_service[`kube-state-metrics` service metrics].
*Kubernetes StatefulSet metrics*:: Monitor StatefulSet configuration, status, and scaling. Learn more at {integrations-docs}/kubernetes/kube-state-metrics#state_statefulset[`kube-state-metrics` StatefulSet metrics].
*Kubernetes StorageClass metrics*:: Monitor how storage is provisioned and allocated. Learn more at {integrations-docs}/kubernetes/kube-state-metrics#state_storageclass[`kube-state-metrics` StorageClass metrics].
====

[discrete]
[[monitor-k8s-kube-state-configure-objects]]
==== Configure the `kube-state-metrics` options

Provide the following information for each of the `kube-state-metrics` options:

[horizontal]

> *Add Metadata*:: Turn on to add metadata to events. The metadata is valuable for grouping and exploring related data.

Reviewer: ```suggestion
*Add Metadata*:: Turn on to add metadata to events. The metadata is valuable for grouping and exploring related data. Default value on
```

. At the bottom of the *Add Kubernetes integration* page, click *Save and continue*.
. Click *Add {agent} to your hosts*.
. Under *Enroll in Fleet?*, select *Run standalone*.
. Under *Configure the agent*, click *Download Manifest*.

After downloading the manifest, update the `ES_USERNAME` and `ES_PASSWORD` environment variables in the DaemonSet to match your {es} credentials.

You can also further modify the manifest to fit your needs. For example, you might want to enable autodiscovery to automatically discover container logs. Refer to the {fleet-guide}/elastic-agent-kubernetes-autodiscovery.html[autodiscovery docs] in the {fleet} guide for more on enabling autodiscovery in your manifest.

Once you are ready to deploy your {agent}:

. From the directory where you've downloaded the manifest, run the following apply command:
+
["source", "sh", subs="attributes"]
----
kubectl apply -f elastic-agent-standalone-kubernetes.yml
----
. Check the {agent} status with the following command:
+
["source", "sh", subs="attributes"]
----
kubectl -n kube-system get pods -l app=elastic-agent

> ----

Reviewer: Refer to https://www.elastic.co/guide/en/fleet/current/debug-standalone-agents.html to debug elastic agent standalone

[discrete]
[[monitor-kubernetes-integration]]
== Part 1: Add and Configure the Kubernetes integration

To start collecting logs and metrics from your Kubernetes clusters, first add the {integrations-docs}/kubernetes[Kubernetes integration] to your policy and configure the metrics and logs you want to collect.

[discrete]
[[monitor-k8s-add-integration]]
=== Step 1: Add the Kubernetes integration to your deployment

Follow these steps to add the Kubernetes integration to your policy:
//add screenshots if necessary

. Select *Add integrations* from your deployment's homepage.
+
[role="screenshot"]
image::images/add-integration.png[Add integration button]
. Enter "Kubernetes" in the search bar, and select the *Kubernetes* integration.
. Click *Add Kubernetes* at the top of the Kubernetes integration page.
. Click *Add integration only (skip agent installation)* at the bottom of the Add integration page.
+
[role="screenshot"]

> image::images/skip-agent-button.png[Add integration only (skip agent installation) button]

Reviewer 2: I'm okay with leaving these as-is, but I do wonder if these screenshots are necessary.

Continue to <<monitor-k8s-configure-integration>>.

[discrete]
[[monitor-k8s-configure-integration]]
=== Step 2: Configure your Kubernetes integration

The Kubernetes integration can fetch metrics and logs from several components.
From the *Add Kubernetes integration* page, configure the options for the metrics and logs you want to collect.
The following sections outline the configuration options.

*Collect Metrics*

Collecting metrics about Kubernetes clusters and the workloads running on top of them is a key aspect of Kubernetes observability.
You need to collect metrics about the resources running on the underlying infrastructure (for example, the machines of your Kubernetes cluster), metrics for the whole cluster, as well as individual metrics for the containers and pods.
Specifically, you need to monitor the health and performance of:

* The hosts where Kubernetes components are running. Each host produces metrics
like CPU, memory, disk utilization, and disk and network I/O.

* Kubernetes containers, which produce their own set of metrics.

* The applications running as Kubernetes pods, such as application servers and
databases, each producing its own set of metrics.

> * Additional Kubernetes resources (like services, deployments, cronjobs) are valuable assets of the whole infrastructure and produce their own set of metrics that need monitoring

Reviewer 2: ```suggestion
* Additional Kubernetes resources (like services, deployments, cronjobs) are valuable assets of the whole infrastructure and produce their own set of metrics that need monitoring.
```

*Bearer Token File*:: The file path to the token used to authenticate with the kubelet API.
*Hosts*:: The address of the kubelet API server that Elastic will connect to for collecting metrics. `${env.NODE_NAME}` is an environment variable that represents the name of the Kubernetes node. Port `10250` is the default port where the kubelet API listens for HTTPS connections.
*Period*:: How frequently to poll the kubelet API for metrics. The default is every 10 seconds.
*SSL Verification Mode*:: Specify how to handle SSL verification. `none` means that SSL verification is disabled.

[discrete]
[[monitor-k8s-kubelet-configure-metadata]]
==== Configure kubelet API metadata enrichment

Under **Kubernetes Container metrics** and **Kubernetes Pod metrics**, you can configure metadata enrichment from **Advanced options → Add node and namespace metadata**.

From here, update the `add_resource_metadata` block to configure enrichment:

["source", "yaml", subs="attributes"]
----
# add_resource_metadata:
#   namespace:
#     include_labels: ["namespacelabel1"]
#   node:
#     include_labels: ["nodelabel2"]
#     include_annotations: ["nodeannotation1"]

> #   deployment: false

Reviewer 2: Should this code be commented out?

*Kubernetes Service metrics*:: Monitor service configuration, accessibility, and network integration. Learn more at {integrations-docs}/kubernetes/kube-state-metrics#state_service[`kube-state-metrics` service metrics].
*Kubernetes StatefulSet metrics*:: Monitor StatefulSet configuration, status, and scaling. Learn more at {integrations-docs}/kubernetes/kube-state-metrics#state_statefulset[`kube-state-metrics` StatefulSet metrics].
*Kubernetes StorageClass metrics*:: Monitor how storage is provisioned and allocated. Learn more at {integrations-docs}/kubernetes/kube-state-metrics#state_storageclass[`kube-state-metrics` StorageClass metrics].
====

[discrete]
[[monitor-k8s-kube-state-configure-objects]]
==== Configure the `kube-state-metrics` options

Provide the following information for each of the `kube-state-metrics` options:

[horizontal]
*Add Metadata*:: This option adds metadata to events. The metadata is valuable for grouping and exploring related data. This option is on by default.
*Hosts*:: The address where `kube-state-metrics` is running. Port 8080 is the default.

> *Leader Election*:: When on (default behavior), only the elastic agent which holds the leadership lock will retrieve metrics from the `kube_state_metrics`. This prevents duplicate data in multi-node Kubernetes clusters.

Reviewer 2: "holds the leadership lock" is a restrictive clause, so you should use `that` instead of `which`.

```suggestion
*Leader Election*:: When on (default behavior), only the Elastic Agent that holds the leadership lock will retrieve metrics from the `kube_state_metrics`. This prevents duplicate data in multi-node Kubernetes clusters.
```

  node:
   enabled: true
    #use_regex_include: false
    include_labels: ["nodelabel2"]
    include_annotations: ["nodeannotation1"]
    #use_regex_exclude: false
    #exclude_labels: ["nodelabel3"]
  #deployment: false
  #cronjob: false
----

[discrete]
[[monitor-kubernetes-integration-k8s-api-metrics]]
=== Collect Kubernetes metrics from Kubernetes API server

Collecting metrics from the `kube-apiserver` is on by default.
The `kube-apiserver` sets up and validates pods, services, and other API objects.
These metrics provide insight into the API server's performance, workload, and health.

Refer to {integrations-docs}/kubernetes/kube-apiserver[`kube-apiserver` metrics] for more on the metrics collected.

[discrete]
[[monitor-k8s-apiserver-configure]]
==== Configure Kubernetes API server options

Provide the following information to collect `kube-apiserver` metrics:

[horizontal]
*Bearer Token File*:: The file path to the token used to authenticate with the `kube-apiserver`.
*Hosts*:: The address of the Kubernetes API server that the integration connects to. It uses the `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` environment variables.

> *Leader Election*:: When on (default behavior) only the elastic agent which holds the leadership lock will retrieve metrics from the `kube-apiserver`. This prevents duplicate data in multi-node Kubernetes clusters.

Reviewer 2: ```suggestion
*Leader Election*:: When on (default behavior), only the Elastic Agent that holds the leadership lock will retrieve metrics from the `kube-apiserver`. This prevents duplicate data in multi-node Kubernetes clusters.
```

> *Period*:: How frequently Elastic to poll the `kube-state-metrics` for metrics. The default is every 30 seconds.

Reviewer 2: ```suggestion
*Period*:: How frequently to poll the `kube-state-metrics` for metrics. The default is every 30 seconds.
```

*SSL Verification Mode*:: Specify how to handle SSL verification. Defaults to `none` meaning that SSL verification is disabled.

[discrete]
[[monitor-kubernetes-integration-k8s-api-events]]
=== Collect Kubernetes events from Kubernetes API Server

Event metrics give you an overall view of what's happening in a cluster.
These metrics help you understand what's happening in your cluster and improve reliability and stability.
Collecting Kubernetes events from the Kubernetes API server is on by default.

Refer to {integrations-docs}/kubernetes/events[events metrics] for more on the metrics collected.

[discrete]
[[monitor-k8s-api-events-configure]]
==== Configure events from the Kubernetes API server

Provide the following information to collect Kubernetes events metrics:

[horizontal]
*Period*:: How frequently Elastic to poll the `kube-api-server` for events. The default is every 10 seconds.
*Add Metadata*:: Turn on to add metadata to events. The metadata is valuable for grouping and exploring related data.
*Skip older events*:: Ignores events that occurred before a certain time

> *Leader Election*:: When on (default behavior) only the elastic agent which holds the leadership lock will retrieve metrics from the `kube-apiserver`. This prevents duplicate data in multi-node Kubernetes clusters.

Reviewer 2: ```suggestion
*Leader Election*:: When on (default behavior), only the Elastic Agent that holds the leadership lock will retrieve metrics from the `kube-apiserver`. This prevents duplicate data in multi-node Kubernetes clusters.
```
